build.gradle

```diff
@@ -76,4 +76,140 @@ project(':configurator') {
+                fileName: 'OpenJDK8U-jre_x64_windows_hotspot_8u212b03.zip'),
+        linux: new JreProperties(hash: '74daf0b77a7fd679cbb3a6228e0efa8c4a90b7664aa057f211e34bbfb38640fb',
+                fileName: 'OpenJDK8U-jre_x64_linux_hotspot_8u212b03.tar.gz'),
+        mac: new JreProperties(hash: '80855320c42a3b06617c6e466c64df67731542a805972185b075ca6cb1222c7f',
+                fileName: 'OpenJDK8U-jre_x64_mac_hotspot_8u212b03.tar.gz')
+]
+
+import java.security.MessageDigest
+
+task downloadJre(description: 'Download OpenJRE of Linux, Windows and Mac, and check their hashes') {
+    doFirst {
+        if (!jreDownloadDir.exists()) {
+            jreDownloadDir.mkdirs()
+        }
+
+        // Filter out entries that already have JRE downloaded
+        jreProperties.findAll {String sys, JreProperties props -> !props.jre.exists()}
+                .collect {String sys, JreProperties props ->
+            Thread.start {
+                println("Start downloading OpenJRE for ${sys}...")
+                new URL(props.downloadUrl as String).withInputStream {
+                    props.jre.newOutputStream() << it
+                }
+                if (checkHash(props)) {
+                    println("OpenJRE for ${sys} is downloaded.")
+                } else {
+                    println("Hash mismatch for downloaded ${sys} JRE. File may have been modified.")
+                    System.exit(1)
+                }
+            }
+        }*.join()
+    }
+}
+
+// Execute downloadJre only when any one of three JREs is missing
+downloadJre.onlyIf {
+    jreProperties.any {sys,props ->
+        !props.jre.exists()
+    }
+}
+
+static boolean checkHash(JreProperties props) {
+    final hash = MessageDigest.getInstance('SHA-256').with {
+        props.jre.eachByte(1024) {buff,len ->
+            update(buff, 0, len)
+        }
+        digest().encodeHex() as String
+    }
+
+    hash == props.hash
+}
+
+task unzipJre(description: 'Unzip a JRE when the extract directory is missing', dependsOn: downloadJre) {
+    doFirst {
+
+        jreProperties.each {sys, props ->
+            // If any one of three JRE folders is missing then only unzip the corresponding JRE
+            if (!new File("${jreDownloadDir.absolutePath}/jre/${sys}").exists()) {
+                if (props.jre.name.endsWith(ZIP)) {
+                    copy {
+                        from zipTree(props.jre)
+                        into "$jreExtractDir/${WINDOWS}"
+                    }
+                } else {
+                    copy {
+                        from tarTree(resources.gzip(props.jre))
+                        into "$jreExtractDir/${props.jre.name.contains(MAC) ? MAC : LINUX}"
+                    }
+                }
+            }
+        }
+    }
+}
+
+build.dependsOn(unzipJre)
+
+distributions {
+    packageProperties.each { String sys, PackageProperties props ->
+        String scriptDir = "${projectDir}/launcher-scripts/${props.scriptFile}"
```

Comment:
Due to the use of Distribution, the only useful field of PackageProperties is scriptFile. For the sake of simplicity, I wonder if we still want to keep PackageProperties, or to just have a simple map like ['windows': 'windows-launcher.bat', 'mac': 'mac-launcher.sh', 'linxu': 'linxu-launcher.sh'].

Reply:
Nah totally, if you only need one value, go with a simple map. Possibly change name though, maybe:

```groovy
final launcherScripts = [
    'windows': 'windows-launcher.bat',
    'linux': etc...
]
```